networks/2b_community_detection.py

```diff
@@ -0,0 +1,202 @@
+
+# %%time
+if method == "fastgreedy":
+    # Simplify graph to remove multiple edges and loops
+    if not pao1_G.is_simple():
+        pao1_G.simplify()
+    if not pa14_G.is_simple():
+        pa14_G.simplify()
+
+    assert pao1_G.is_simple()
+    assert pa14_G.is_simple()
+
+    # Detection method
+    pao1_partition = pao1_G.community_fastgreedy(weights=pao1_G.es["weight"])
+    pa14_partition = pa14_G.community_fastgreedy(weights=pa14_G.es["weight"])
+
+# +
+# pao1_G.vs.attribute_names()
+# -
+
+# ### Walktrap
+# This algorithm performs random walks using a specified step size. Where densely connected areas occur, the random walk becomes “trapped” in local regions that then define communities
+#
+# Returns VertextDendrogram
+
+# %%time
+if method == "walktrap":
+    pao1_partition = pao1_G.community_walktrap(
+        weights=pao1_G.es["weight"], steps=nsteps
+    )
+    pa14_partition = pa14_G.community_walktrap(
+        weights=pa14_G.es["weight"], steps=nsteps
+    )
+
+# ### Multilevel
+# This algorithm is similar to fastgreedy, but it merges communities to optimize modularity based upon only the neighboring communities as opposed to all communities. The algorithm terminates when only a single node is left, or when the improvement in modularity cannot result from the simple merge of two neighboring communities. (Louvain clustering)
+#
+# Returns VertexClustering
+
+# %%time
+if method == "louvain":
+    pao1_partition = pao1_G.community_multilevel(
+        weights=pao1_G.es["weight"], return_levels=False
+    )
+    pa14_partition = pa14_G.community_multilevel(
+        weights=pa14_G.es["weight"], return_levels=False
+    )
+
+# ### Infomap
+# This algorithm uses the probability flow of information in random walks, which occurs more readily in groups of heavily connected nodes. Thus, information about network structure can be compressed in maps of modules (nodes where information travels quickly)
+#
+# Returns VertexClustering
+
+# %%time
+if method == "infomap":
+    pao1_partition = pao1_G.community_infomap(
+        edge_weights=pao1_G.es["weight"], trials=ntrials
+    )
+    pa14_partition = pa14_G.community_infomap(
+        edge_weights=pa14_G.es["weight"], trials=ntrials
+    )
+
+
+# ## Get membership
+
+# get dataframe mapping Pa genes to communities
+def graph_partition_to_df(G, partition, method):
+    if method in ["louvain", "infomap"]:
+        clusters = []
+        for label, vl in enumerate(partition):
+            clusters += [(G.vs["name"][v], label, G.degree(v)) for v in vl]
+
+        membership_df = pd.DataFrame(clusters, columns=["gene", "module id", "degree"])
+        membership_df = membership_df.set_index("gene")
+
+        return membership_df
+
+
+# +
+# pao1_partition.es.attribute_names()
+# -
+
+pao1_membership_df = graph_partition_to_df(pao1_G, pao1_partition, method)
+print(len(pao1_membership_df["module id"].unique()))
+pao1_membership_df.sort_values(by="degree", ascending=False).head()
```

Comment:
I think responds to a previous question.

Reply:
Here "module id" can be read as the label, right?

Reply:
Yes that is correct